libs/core/threading_base/include/hpx/threading_base/annotated_function.hpp

```diff
@@ -140,6 +140,7 @@ namespace hpx { namespace util {
                 Ts...>::type
             operator()(Ts&&... ts)
             {
+                annotate_function annotate(get_function_annotation());
```

Comment:
I *think* the reason there was no `annotate_function` here is that typically the annotation of an `annotated_function` will already be applied when the task is spawned, meaning that using `annotate_function` here would create a new "task" in the trace with the same name. Leaving it out here of course means that just calling `operator()` on an `annotated_function` not as a task function will not apply the annotation.

Could `annotate_function` only apply the new annotation if it's different (either simply comparing the pointers, or actually comparing the strings)? Maybe it already does that?

Reply:
> I _think_ the reason there was no `annotate_function` here is that typically the annotation of an `annotated_function` will already be applied when the task is spawned, meaning that using `annotate_function` here would create a new "task" in the trace with the same name. Leaving it out here of course means that just calling `operator()` on an `annotated_function` not as a task function will not apply the annotation.

Yes, that was the intend. However, IIRC, `annotated_function` had all kind of problems in the past that APEX didn't properly pick up the annotation. Adding the `annotate_function` in the function operator should fix this. 

> Could `annotate_function` only apply the new annotation if it's different (either simply comparing the pointers, or actually comparing the strings)? Maybe it already does that?

I can certainly do this.

Reply:
Thanks! Yep, agreed, it was kind of odd before so probably applying the annotation (even if it's unconditional) is probably the lesser evil.